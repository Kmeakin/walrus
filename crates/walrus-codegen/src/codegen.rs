#![allow(clippy::cast_possible_truncation)]

use crate::free_vars::FreeVars;
use arena::ArenaMap;
use either::{self, *};
pub use inkwell::context::Context;
use inkwell::{
    basic_block::BasicBlock,
    builder::Builder,
    memory_buffer::MemoryBuffer,
    module::Module,
    targets::TargetData,
    types::{AnyType, BasicType, BasicTypeEnum, FunctionType, IntType, StructType},
    values::{BasicValue, BasicValueEnum, FunctionValue, IntValue, PointerValue},
    AddressSpace, FloatPredicate, IntPredicate,
};
use std::{collections::HashMap, ops::Index};
use walrus_semantics::{
    builtins::Builtin,
    hir::{
        self, ArithmeticBinop, Binop, CmpBinop, EnumDefId, Expr, ExprId, Field, FieldInit, FnDefId,
        LazyBinop, Lit, MatchCase, Param, Pat, PatId, StructDefId, StructField, Unop, VarId,
    },
    scopes::{self, Denotation},
    ty,
    ty::{FnType, PrimitiveType, Type},
};

mod expr;
mod pat;
mod trivial;
mod types;
mod util;

pub struct Compiler<'ctx> {
    pub llvm: &'ctx Context,
    pub module: Module<'ctx>,
    pub builder: Builder<'ctx>,

    pub hir: hir::HirData,
    pub scopes: scopes::Scopes,
    pub types: ty::InferenceResult,
}

#[derive(Debug, Clone)]
pub struct Loop<'ctx> {
    body_bb: BasicBlock<'ctx>,
    exit_bb: BasicBlock<'ctx>,
    result_alloca: PointerValue<'ctx>,

    does_continue: bool,
    does_break: bool,
}

#[derive(Debug, Clone, Default)]
pub struct Vars<'a> {
    locals: ArenaMap<VarId, PointerValue<'a>>,
    fns: ArenaMap<FnDefId, FunctionValue<'a>>,
    types: HashMap<Either<StructDefId, EnumDefId>, StructType<'a>>,
    string_type: Option<StructType<'a>>,
    current_loop: Option<Loop<'a>>,
}

impl<'a> Index<VarId> for Vars<'a> {
    type Output = PointerValue<'a>;
    fn index(&self, id: VarId) -> &Self::Output { &self.locals[id] }
}
impl<'a> Index<FnDefId> for Vars<'a> {
    type Output = FunctionValue<'a>;
    fn index(&self, id: FnDefId) -> &Self::Output { &self.fns[id] }
}

type Value<'ctx> = Option<BasicValueEnum<'ctx>>;

impl<'ctx> Compiler<'ctx> {
    #[allow(clippy::useless_transmute)]
    pub fn codegen_module(self) -> String {
        let this: &'static Compiler<'static> = unsafe { std::mem::transmute(&self) };

        let builtins_source = include_str!("builtins.ll");
        let builtins =
            MemoryBuffer::create_from_memory_range_copy(builtins_source.as_bytes(), "builtins");
        let builtins = this
            .llvm
            .create_module_from_ir(builtins)
            .map_err(|e| eprintln!("{}", e.to_string()))
            .unwrap();
        this.module.link_in_module(builtins).unwrap();
        let mut vars = Vars::default();

        for (id, func) in this.hir.fn_defs.iter() {
            let fn_type = this.known_fn_type(&mut vars, &this.types[id]);
            let name = this.hir[func.name].as_str();
            let llvm_fn = this.module.add_function(name, fn_type, None);
            vars.fns.insert(id, llvm_fn);
        }

        for (id, _) in this.hir.fn_defs.iter() {
            this.codegen_fn(&mut vars, id)
        }

        if let Err(e) = this.module.verify() {
            eprintln!("{}", this.module.print_to_string().to_string());
            eprintln!("{}", e.to_string());
            panic!()
        }

        this.module.print_to_string().to_string()
    }

    fn codegen_fn(&'ctx self, vars: &mut Vars<'ctx>, id: FnDefId) {
        let llvm_fn = vars[id];
        let fn_def = &self.hir[id];
        let name = &self.hir[fn_def.name];
        let bb = self
            .llvm
            .append_basic_block(llvm_fn, &format!("{name}.entry"));
        self.builder.position_at_end(bb);

        llvm_fn
            .get_param_iter()
            .zip(&fn_def.params)
            .enumerate()
            .for_each(|(idx, (llvm_param, hir_param))| {
                llvm_param.set_name(&format!("{name}.params.{idx}"));
                self.codegen_pat(vars, hir_param.pat, llvm_param)
            });

        if let Some(value) = self.codegen_expr(vars, fn_def.expr) {
            self.builder.build_return(Some(&value));
        }
    }
}
