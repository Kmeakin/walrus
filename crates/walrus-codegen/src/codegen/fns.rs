use super::*;

/// Wrapping global functions
impl<'ctx> Compiler<'ctx> {
    /// Wrap a `FunctionValue` in a closure with an empty environment
    pub fn codegen_fn_wrapper(
        &self,
        vars: &mut Vars<'ctx>,
        fn_name: &str,
        fn_value: FunctionValue<'ctx>,
        fn_type: &FnType,
    ) -> BasicValueEnum {
        let code_ptr = fn_value.as_global_value().as_pointer_value();
        let code_ptr = self.builder.build_bitcast(
            code_ptr,
            self.fn_type(vars, fn_type).ptr_type(AddressSpace::Generic),
            &format!("{fn_name}.closure"),
        );
        let closure = self
            .closure_type(vars, fn_type)
            .const_named_struct(&[code_ptr, self.codegen_null_ptr()]);
        closure.into()
    }

    /// Wrap a builtin function in a closure with an empty environment
    pub fn codegen_builtin_fn_wrapper(
        &self,
        vars: &mut Vars<'ctx>,
        builtin: Builtin,
    ) -> BasicValueEnum {
        match builtin {
            Builtin::Fn { name, ty } => {
                let fn_value = self.get_builtin_fn(vars, name, &ty);
                self.codegen_fn_wrapper(vars, name, fn_value, &ty)
            }
            Builtin::Type { .. } => {
                unreachable!("Attempt to codegen non-value builtin: {:#?}", builtin)
            }
        }
    }

    fn get_builtin_fn(
        &self,
        vars: &mut Vars<'ctx>,
        name: &str,
        fn_type: &FnType,
    ) -> FunctionValue<'ctx> {
        match vars.builtin_fns.get(name) {
            Some(fn_value) => *fn_value,
            None => {
                let fn_value = self.module.add_function(
                    &format!("builtin_{name}"),
                    self.known_fn_type(vars, fn_type),
                    None,
                );
                vars.builtin_fns.insert(name.to_string(), fn_value);
                fn_value
            }
        }
    }
}

/// Calling functions
impl<'ctx> Compiler<'ctx> {
    fn build_call(
        &'ctx self,
        fn_value: impl Into<Either<FunctionValue<'ctx>, PointerValue<'ctx>>>,
        args: &[Value<'ctx>],
        fn_name: &str,
    ) -> OptValue<'ctx> {
        let call = self
            .builder
            .build_call(fn_value.into(), args, &format!("{fn_name}.call"))
            .try_as_basic_value();
        match call {
            Left(value) => Some(value),
            Right(_) => None,
        }
    }

    pub fn codegen_fn_args(
        &'ctx self,
        vars: &mut Vars<'ctx>,
        args: &[ExprId],
    ) -> Option<Vec<Value>> {
        args.iter()
            .map(|arg| self.codegen_expr(vars, *arg))
            .collect::<Option<Vec<_>>>()
    }

    fn codegen_known_fn_call(
        &'ctx self,
        vars: &mut Vars<'ctx>,
        fn_id: FnDefId,
        args: &[Value<'ctx>],
    ) -> OptValue<'ctx> {
        let fn_def = &self.hir[fn_id];
        let fn_name = &self.hir[fn_def.name].as_str();
        let fn_value = vars[fn_id];
        self.build_call(fn_value, args, fn_name)
    }

    pub fn codegen_builtin_fn_call(
        &'ctx self,
        vars: &mut Vars<'ctx>,
        name: &str,
        fn_type: &FnType,
        args: &[Value<'ctx>],
    ) -> OptValue<'ctx> {
        let fn_value = self.get_builtin_fn(vars, name, fn_type);
        self.build_call(fn_value, args, name)
    }

    fn codegen_lambda_call(
        &'ctx self,
        vars: &mut Vars<'ctx>,
        func: ExprId,
        args: &[Value<'ctx>],
    ) -> OptValue<'ctx> {
        let mut args = args.to_vec();
        let closure_value = self.codegen_expr(vars, func)?.into_struct_value();
        let code_ptr = self
            .builder
            .build_extract_value(closure_value, 0, "closure.code")
            .unwrap()
            .into_pointer_value();
        let env_ptr = self
            .builder
            .build_extract_value(closure_value, 1, "closure.env")
            .unwrap();
        args.insert(0, env_ptr);
        self.build_call(code_ptr, &args, "lambda")
    }

    pub fn codegen_call(
        &'ctx self,
        vars: &mut Vars<'ctx>,
        func: ExprId,
        args: &[Value<'ctx>],
    ) -> OptValue<'ctx> {
        let func_val = &self.hir[func];
        match func_val {
            Expr::Var(var_id) => {
                let var = &self.hir[*var_id];
                let denotation = self.scopes.lookup_var(*var_id, var);
                match denotation {
                    Some(Denotation::Fn(id)) => self.codegen_known_fn_call(vars, id, args),
                    Some(Denotation::Builtin(Builtin::Fn { name, ty })) => {
                        self.codegen_builtin_fn_call(vars, name, &ty, args)
                    }
                    _ => self.codegen_lambda_call(vars, func, args),
                }
            }
            _ => self.codegen_lambda_call(vars, func, args),
        }
    }
}

/// Constructing lambdas
impl<'ctx> Compiler<'ctx> {
    pub fn codegen_lambda(
        &'ctx self,
        vars: &mut Vars<'ctx>,
        expr: ExprId,
        params: &[Param],
        body: ExprId,
    ) -> BasicValueEnum {
        let free_vars = self.free_vars(expr);

        let code_ptr = self
            .codegen_lambda_body(
                // this clone is necessary. Dont delete it!
                &mut vars.clone(),
                &free_vars,
                expr,
                params,
                body,
            )
            .as_global_value()
            .as_pointer_value();

        let closure_type = self.types[expr].as_fn().unwrap();
        let closure_alloca = self
            .builder
            .build_alloca(self.closure_type(vars, closure_type), "closure.alloca");
        let code_gep = self
            .builder
            .build_struct_gep(closure_alloca, 0, "closure.code")
            .unwrap();
        self.builder.build_store(code_gep, code_ptr);

        let env_gep = self
            .builder
            .build_struct_gep(closure_alloca, 1, "closure.env")
            .unwrap();

        // store free variables
        let free_vars_type = self.tuple_type(
            vars,
            &free_vars
                .iter()
                .map(|(pat, _)| self.types[pat].clone())
                .collect::<Vec<_>>(),
        );
        let env_alloca = self
            .builder
            .build_malloc(free_vars_type, "env.malloc")
            .unwrap();
        for (idx, (free_var, ())) in free_vars.iter().enumerate() {
            let gep = self
                .builder
                .build_struct_gep(env_alloca, idx as u32, &format!("env.{idx}.gep"))
                .unwrap();
            let val = self
                .builder
                .build_load(vars[free_var], &format!("env.{idx}"));
            self.builder.build_store(gep, val);
        }
        let env_alloca = self
            .builder
            .build_bitcast(env_alloca, self.void_ptr_type(), "env");
        self.builder.build_store(env_gep, env_alloca);
        self.builder.build_load(closure_alloca, "closure")
    }

    fn codegen_lambda_body(
        &'ctx self,
        vars: &mut Vars<'ctx>,
        free_vars: &FreeVars,
        expr: ExprId,
        params: &[Param],
        body: ExprId,
    ) -> FunctionValue<'ctx> {
        let fn_type = self.fn_type(vars, self.types[expr].as_fn().unwrap());
        let llvm_fn = self.module.add_function("lambda", fn_type, None);
        let old_bb = self.builder.get_insert_block().unwrap();

        let bb = self.llvm.append_basic_block(llvm_fn, "lambda.entry");
        self.builder.position_at_end(bb);

        // load free vars
        let free_vars_type = self.tuple_type(
            vars,
            &free_vars
                .iter()
                .map(|(pat, _)| self.types[pat].clone())
                .collect::<Vec<_>>(),
        );

        let env_param = llvm_fn.get_nth_param(0).unwrap();
        env_param.set_name("env_ptr");
        let env_ptr = self.builder.build_bitcast(
            env_param,
            free_vars_type.ptr_type(AddressSpace::Generic),
            "env_ptr",
        );
        let env = self.builder.build_load(env_ptr.into_pointer_value(), "env");
        for (idx, (free_var, ())) in free_vars.iter().enumerate() {
            let val = self
                .builder
                .build_extract_value(env.into_struct_value(), idx as u32, &format!("env.{idx}"))
                .unwrap();
            self.codegen_local_var(vars, free_var, val)
        }

        // load params
        llvm_fn
            .get_param_iter()
            .skip(1)
            .zip(params)
            .enumerate()
            .for_each(|(idx, (llvm_param, hir_param))| {
                llvm_param.set_name(&format!("params.{idx}"));
                self.codegen_pat(vars, hir_param.pat, llvm_param)
            });

        if let Some(value) = self.codegen_expr(vars, body) {
            self.builder.build_return(Some(&value));
        }

        self.builder.position_at_end(old_bb);

        llvm_fn
    }
}
