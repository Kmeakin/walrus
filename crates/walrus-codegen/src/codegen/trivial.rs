use super::*;
use inkwell::values::FloatValue;

impl<'ctx> Compiler<'ctx> {
    pub fn codegen_unit(&self) -> BasicValueEnum { self.llvm.const_struct(&[], false).into() }

    pub fn codegen_null_ptr(&self) -> BasicValueEnum { self.void_ptr_type().const_zero() }

    pub fn codegen_undef(&self) -> BasicValueEnum { self.llvm.i8_type().get_undef().into() }

    pub fn codegen_lit(&self, vars: &mut Vars<'ctx>, lit: &Lit) -> BasicValueEnum<'ctx> {
        match lit {
            Lit::Bool(b) => self.llvm.bool_type().const_int(*b as _, false).into(),
            Lit::Int(val) => self.llvm.i32_type().const_int((*val).into(), false).into(),
            Lit::Float(val) => self.llvm.f32_type().const_float((*val).into()).into(),
            Lit::Char(val) => self.llvm.i32_type().const_int((*val).into(), false).into(),
            Lit::String(val) => self.codegen_string(vars, val),
        }
    }

    pub fn codegen_bool(&self, val: bool) -> IntValue<'ctx> {
        self.llvm.bool_type().const_int(val as _, false)
    }

    pub fn codegen_true(&self) -> IntValue<'ctx> { self.codegen_bool(true) }
    pub fn codegen_false(&self) -> IntValue<'ctx> { self.codegen_bool(false) }

    pub fn codegen_int(&self, val: u32) -> IntValue<'ctx> {
        self.llvm.i32_type().const_int(val.into(), false)
    }

    pub fn codegen_float(&self, val: f32) -> FloatValue<'ctx> {
        self.llvm.f32_type().const_float(val.into())
    }

    pub fn codegen_char(&self, val: char) -> IntValue<'ctx> {
        self.llvm.i32_type().const_int(val as _, false)
    }

    pub fn codegen_string(
        &self,
        vars: &mut Vars<'ctx>,
        val: &smol_str::SmolStr,
    ) -> BasicValueEnum<'ctx> {
        let len = self.codegen_int(val.len() as _);

        // TODO: this truncates the string at the first null byte
        let bytes = self.builder.build_global_string_ptr(val, "string");
        self.string_type(vars)
            .const_named_struct(&[len.into(), bytes.as_pointer_value().into()])
            .into()
    }
}
