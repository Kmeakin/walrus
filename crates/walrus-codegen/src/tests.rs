use crate::codegen::Context;
use inkwell::{memory_buffer::MemoryBuffer, OptimizationLevel};
use insta::*;

#[track_caller]
fn test_codegen(src: &str) { test::<()>(src, None) }

#[track_caller]
fn test_codegen_and_run<T>(src: &str, expected: T)
where
    T: PartialEq + std::fmt::Debug,
{
    test(src, Some(expected))
}

fn test<T>(src: &str, expected: Option<T>)
where
    T: PartialEq + std::fmt::Debug,
{
    let syntax = walrus_parser::parse(src);
    let hir = walrus_semantics::hir::lower(&syntax);
    let scopes = walrus_semantics::scopes::scopes(&hir);
    let types = walrus_semantics::ty::infer(hir.clone(), scopes.clone());

    dbg!(&syntax);
    dbg!(&hir);
    dbg!(&scopes);
    dbg!(&types);

    let module = crate::codegen("module", hir.hir, scopes, types);

    let mut settings = insta::Settings::new();
    settings.set_snapshot_path("../snapshots");
    settings.set_prepend_module_to_snapshot(false);
    settings.bind(|| assert_display_snapshot!(module));

    let llvm = Context::create();
    let module_buffer = MemoryBuffer::create_from_memory_range_copy(module.as_bytes(), "module");
    let llvm_module = llvm.create_module_from_ir(module_buffer).unwrap();

    if let Some(expected) = expected {
        let exec_engine = llvm_module
            .create_jit_execution_engine(OptimizationLevel::None)
            .unwrap();
        let f = unsafe { exec_engine.get_function::<unsafe extern "C" fn() -> T>("main") }.unwrap();
        assert_eq!(unsafe { f.call() }, expected);
    }
}

mod literals {
    use super::*;
    #[test]
    fn empty_fn() { test_codegen_and_run("fn main() -> _ {}", ()) }

    #[test]
    fn lit_true() { test_codegen_and_run(r#"fn main() -> _ { true }"#, true) }

    #[test]
    fn lit_false() { test_codegen_and_run(r#"fn main() -> _ { false }"#, false) }

    #[test]
    fn lit_int() { test_codegen_and_run(r#"fn main() -> _ { 1 }"#, 1_i32) }

    #[test]
    fn lit_float() { test_codegen_and_run(r#"fn main() -> _ { 1.234 }"#, 1.234_f32) }

    #[test]
    fn lit_char() { test_codegen_and_run(r#"fn main() -> _ { 'a' }"#, 'a') }

    #[test]
    fn lit_string() {
        test_codegen_and_run(r#"fn main() -> _ { "hello world" }"#, "hello world".len())
    }
}

mod tuples {
    use super::*;
    #[test]
    fn tuple0() { test_codegen_and_run(r#"fn main() -> _ { () }"#, ()) }

    #[test]
    fn tuple1() { test_codegen_and_run(r#"fn main() -> _ { (1,) }"#, (1_i32,)) }

    #[test]
    fn tuple2() { test_codegen_and_run(r#"fn main() -> _ { (1, 2) }"#, (1_i32, 2_i64)) }
}

mod if_ {
    use super::*;
    #[test]
    fn if_then_else_true() {
        test_codegen_and_run("fn main() -> _ { if true {1} else {0} }", 1_i32)
    }

    #[test]
    fn if_then_else_false() {
        test_codegen_and_run("fn main() -> _ { if false {1} else {0} }", 0_i32)
    }

    #[test]
    fn if_then_true() { test_codegen_and_run(r#"fn main() -> _ { if true {1} }"#, ()); }
    #[test]
    fn if_then_false() { test_codegen_and_run(r#"fn main() -> _ { if false {1} }"#, ()); }
}

mod let_ {
    use super::*;

    #[test]
    fn let_var() { test_codegen_and_run(r#"fn main() -> _ { let x = 5; x }"#, 5_i32); }

    #[test]
    fn let_tuple2() { test_codegen_and_run(r#"fn main() -> _ { let (x, y) = (5, 6); x }"#, 5_i32) }

    #[test]
    fn let_tuple4() {
        test_codegen_and_run(
            r#"
fn main() -> _ {
    let (a, b, c, d) = (9, 8, 7, 6);
    let (x, (y, z), w) = (d, (c, b), a);
    x
}"#,
            6_i32,
        )
    }
}

mod fns {
    use super::*;
    #[test]
    fn id_fn() {
        test_codegen_and_run(
            r#"
fn main() -> _ { id(5) }
fn id(x: _) -> _ {x}
"#,
            5_i32,
        )
    }

    #[test]
    fn get_five_fn() {
        test_codegen_and_run(
            r#"
fn main() -> _ { get_five() }
fn get_five() -> _ {5}
"#,
            5_i64,
        )
    }

    #[test]
    fn fn_value() {
        test_codegen_and_run(
            r#"
fn main() -> _ { 
    let f = get_five;
    f()
}

fn get_five() -> _ {5}
"#,
            5_i32,
        )
    }

    #[test]
    fn builtin_fn_value() {
        test_codegen_and_run(
            r#"
fn main() -> _ { 
    let f = putchar;
    f('a')
}
"#,
            (),
        )
    }
}

mod return_ {
    use super::*;
    #[test]
    fn early_return() {
        test_codegen_and_run(
            r#"
fn main() -> _ {
    return 5;
    6
}
"#,
            5_i64,
        )
    }

    #[test]
    fn multiple_return() {
        test_codegen_and_run(
            r#"
fn main() -> _ {
    return 5;
    return 6;
}
"#,
            5_i64,
        )
    }

    #[test]
    fn multi_expr_return() { test_codegen_and_run(r#"fn main() -> _ { 1 + (return 2) }"#, 2); }

    #[test]
    #[ignore]
    fn if_then_return() {
        test_codegen_and_run(
            r#"
fn main() -> _ {
    if true {return 1}
    else {0}
}
"#,
            1,
        )
    }
}

mod bool_ops {
    use super::*;
    #[test]
    fn bool_not() { test_codegen_and_run(r#"fn main() -> _  {!false}"#, true) }
    #[test]
    fn bool_or() { test_codegen_and_run(r#"fn main() -> _  {false || true}"#, true) }

    #[test]
    fn bool_and() { test_codegen_and_run(r#"fn main() -> _ {true && false}"#, false) }
}

mod int_ops {
    use super::*;

    #[test]
    fn int_neg() { test_codegen_and_run(r#"fn main() ->   _  {-1}"#, -1_i32) }
    #[test]
    fn int_plus() { test_codegen_and_run(r#"fn main() ->   _  {+1}"#, 1_i32) }

    #[test]
    fn int_add() { test_codegen_and_run(r#"fn main() -> _ {let x = 1; let y = 1; x + y}"#, 2_i32) }
    #[test]
    fn int_sub() { test_codegen_and_run(r#"fn main() -> _ {let x = 1; let y = 1; x - y}"#, 0_i32) }
    #[test]
    fn int_mul() { test_codegen_and_run(r#"fn main() -> _ {let x = 1; let y = 2; x * y}"#, 2_i32) }
    #[test]
    fn int_div() { test_codegen_and_run(r#"fn main() -> _ {let x = 5; let y = 2; x / y}"#, 2_i32) }

    #[test]
    fn int_eq() { test_codegen_and_run(r#"fn main() -> _ {let x = 1; let y = 1; x == y}"#, true) }
    #[test]
    fn int_neq() { test_codegen_and_run(r#"fn main() -> _ {let x = 1; let y = 1; x != y}"#, false) }
    #[test]
    fn int_less() { test_codegen_and_run(r#"fn main() -> _ {let x = 1; let y = 1; x < y}"#, false) }
    #[test]
    fn int_less_eq() {
        test_codegen_and_run(r#"fn main() -> _ {let x = 1; let y = 1; x <= y}"#, true)
    }
    #[test]
    fn int_greater() {
        test_codegen_and_run(r#"fn main() -> _ {let x = 1; let y = 1; x > y}"#, false)
    }
    #[test]
    fn int_greater_eq() {
        test_codegen_and_run(r#"fn main() -> _ {let x = 1; let y = 1; x >= y}"#, true)
    }
}

mod float_ops {
    use super::*;

    #[test]
    fn float_neg() { test_codegen_and_run(r#"fn main() -> _  {-1.0}"#, -1.0_f32) }
    #[test]
    fn float_plus() { test_codegen_and_run(r#"fn main() -> _  {+1.0}"#, 1.0_f32) }

    #[test]
    fn float_add() {
        test_codegen_and_run(
            r#"fn main() -> _ {let x = 1.0; let y = 1.0; x + y}"#,
            2.0_f32,
        )
    }
    #[test]
    fn float_sub() {
        test_codegen_and_run(
            r#"fn main() -> _ {let x = 1.0; let y = 1.0; x - y}"#,
            0.0_f32,
        )
    }
    #[test]
    fn float_mul() {
        test_codegen_and_run(
            r#"fn main() -> _ {let x = 1.0; let y = 2.0; x * y}"#,
            2.0_f32,
        )
    }
    #[test]
    fn float_div() {
        test_codegen_and_run(
            r#"fn main() -> _ {let x = 5.0; let y = 2.0; x / y}"#,
            2.5_f32,
        )
    }

    #[test]
    fn float_eq() {
        test_codegen_and_run(r#"fn main() -> _ {let x = 1.0; let y = 1.0; x == y}"#, true)
    }
    #[test]
    fn float_neq() {
        test_codegen_and_run(
            r#"fn main() -> _ {let x = 1.0; let y = 1.0; x != y}"#,
            false,
        )
    }
    #[test]
    fn float_less() {
        test_codegen_and_run(r#"fn main() -> _ {let x = 1.0; let y = 1.0; x < y}"#, false)
    }
    #[test]
    fn float_less_eq() {
        test_codegen_and_run(r#"fn main() -> _ {let x = 1.0; let y = 1.0; x <= y}"#, true)
    }
    #[test]
    fn float_greater() {
        test_codegen_and_run(r#"fn main() -> _ {let x = 1.0; let y = 1.0; x > y}"#, false)
    }
    #[test]
    fn float_greater_eq() {
        test_codegen_and_run(r#"fn main() -> _ {let x = 1.0; let y = 1.0; x >= y}"#, true)
    }
}

mod assign {
    use super::*;

    #[test]
    fn assign_var() {
        test_codegen_and_run(
            r#"fn main() -> _ {
        let mut x = 5;
        x = 6;
        x
    }"#,
            6_i32,
        )
    }

    #[test]
    fn assign_field() {
        test_codegen_and_run(
            r#"
            struct S {x: Int}

            fn main() -> _ {
                let mut s = S {x: 0};
                s.x = 1;
                s.x
        }"#,
            1_i32,
        )
    }
}

mod builtins {
    use super::*;

    #[test]
    fn builtin_exit() { test_codegen("fn main() -> Int { exit(1) }") }

    #[test]
    fn builtin_putchar() {
        test_codegen_and_run(
            r#"
fn main() -> () {
    putchar('h');
    putchar('e');
    putchar('l');
    putchar('l');
    putchar('o');
    putchar(' ');
    putchar('w');
    putchar('o');
    putchar('r');
    putchar('l');
    putchar('d');
}
"#,
            (),
        )
    }
}

mod loops {
    use super::*;

    // this successfully loops forever, which makes the test never complete!
    #[test]
    fn loop_forever() { test_codegen("fn main() -> Int { loop {} }") }

    // this successfully loops forever, which makes the test never complete!
    #[test]
    fn loop_and_continue() { test_codegen("fn main() -> () { loop {let x = 5; continue} }") }

    #[test]
    fn loop_and_break() {
        test_codegen_and_run("fn main() -> _ { loop {let x = 5; break x} }", 5_i32)
    }

    #[test]
    #[ignore]
    fn loop_and_break_and_continue() {
        test_codegen_and_run(
            "fn main() -> _ { loop { if true {continue} else {break} } }",
            (),
        )
    }
}

mod structs {
    use super::*;
    #[test]
    fn struct_constructor() {
        test_codegen_and_run(
            r#"
struct Foo {x: Int}
fn main() -> _ {
    let foo = Foo {x:5};
    foo
}
"#,
            5_i32,
        )
    }

    #[test]
    fn field_expr() {
        test_codegen_and_run(
            r#"
struct Foo {x: Int,y:Int}
fn main() -> _ {
    let foo = Foo {y:6, x:5};
    foo.y
}
"#,
            6_i32,
        )
    }

    #[test]
    fn struct_pat() {
        test_codegen_and_run(
            r#"
struct Foo {x: Int, y: Int}
fn main() -> _ {
    let Foo {y,x} = Foo{x: 5, y: 10};
    y - x
}
    "#,
            5_i32,
        );
    }

    #[test]
    fn struct_in_struct() {
        test_codegen_and_run(
            r#"
struct Foo {bar: Bar}
struct Bar {x: Int, y: Int}
fn main() -> _ {
    let bar = Bar{x: 5, y: 10};
    let foo = Foo{bar: bar};
    foo.bar.y - foo.bar.x
}
    "#,
            5_i32,
        );
    }
}

mod enums {
    use super::*;

    #[test]
    fn enum0() {
        test_codegen(
            r#"
        enum Void {}
        fn main() -> _ {
            let impossible: Void = exit(0);
            impossible
        }
    "#,
        )
    }

    #[test]
    fn enum1() {
        test_codegen_and_run(
            r#"
        enum Foo { 
            Bar {} 
        }
        fn main() -> _ {
            Foo::Bar{}
        }
    "#,
            ((), ()),
        )
    }

    #[test]
    fn enum1_with_payload() {
        test_codegen_and_run(
            r#"
        enum Foo { 
            Bar {x: Int} 
        }
        fn main() -> _ {
            Foo::Bar{x: 5}
        }
    "#,
            (5_i32, ()),
        )
    }

    #[test]
    fn enum2_variant0() {
        test_codegen_and_run(
            r#"
        enum Foo { 
            Bar {},
            Baz {}
        }
        fn main() -> _ {
            Foo::Bar{}
        }
    "#,
            (0_u8, ()),
        );
    }
    #[test]
    fn enum2_variant1() {
        test_codegen_and_run(
            r#"
        enum Foo { 
            Bar {},
            Baz {}
        }
        fn main() -> _ {
            Foo::Baz{}
        }
    "#,
            (1_u8, ()),
        )
    }

    #[test]
    fn enum2_with_payload_variant0() {
        test_codegen_and_run(
            r#"
        enum Foo { 
            Bar {x: Int},
            Baz {y: Bool} 
        }
        fn main() -> _ {
            Foo::Bar{x: 5}
        }
    "#,
            (0_u8, 5_i64),
        )
    }

    #[test]
    #[ignore]
    fn enum2_with_payload_variant1() {
        test_codegen_and_run(
            r#"
        enum Foo { 
            Bar {x: Int}, 
            Baz {y: Bool} 
        }
        fn main() -> _ {
            Foo::Baz{y: true}
        }
    "#,
            (1_u8, 1_i32),
        )
    }
}

mod pattern_matching {
    use super::*;

    #[test]
    fn simple_pattern_match() {
        test_codegen_and_run(
            r#"
enum Option {
    None {},
    Some {x: Int},
}

fn main() -> _ {
    let opt = Option::Some{x:1};
    match opt {
        Option::None{} => 0,
        Option::Some{x} => x,
    }
}
    "#,
            1_i32,
        )
    }

    #[test]
    fn add_options() {
        test_codegen_and_run(
            r#"
enum Option {
    None {},
    Some {x: Int},
}

fn add_options(a: Option, b: Option) -> Option {
    match (a, b) {
        (Option::Some{x:a}, Option::Some{x:b}) => Option::Some{x: a + b},
        _ => Option::None{},
    }
}

fn main() -> _ {
    let a = Option::Some{x:1};
    let b = Option::Some{x:2};
    add_options(a,b)
}
    "#,
            (1_i8, 3_i64),
        )
    }

    #[test]
    fn list_len() {
        test_codegen_and_run(
            r#"
enum List {
    Nil {},
    Cons {head: Int, tail: List}
}

fn length(l: List) -> Int {
    match l {
        List::Nil {} => 0,
        List::Cons{head: _, tail} => 1 + length(tail),
    }
}

fn main() -> _ {
    let l = List::Cons{head: 2, tail: List::Cons {head: 1, tail: List::Nil{}}};
    length(l)
}
        "#,
            2_i32,
        )
    }

    #[test]
    fn list_map() {
        test_codegen_and_run(
            r#"
enum List {
    Nil {},
    Cons {head: Int, tail: List}
}
 
fn map(l: List, f: (Int) -> Int) -> List {
    match l {
        List::Nil {} => List::Nil{},
        List::Cons{head, tail} => List::Cons{head: f(head), tail: map(tail, f)}
    }
}

fn sum(l: List) -> Int {
    match l {
        List::Nil {} => 0,
        List::Cons{head, tail} => head + sum(tail),
    }
}

fn main() -> _ {
    let l1 = List::Cons{head: 4, tail: List::Cons{head: 3, tail: List::Cons{head: 2, tail: List::Cons{head: 1, tail: List::Nil {}}}}};
    let l2 = map(l1, (x) => x * x);
    let s = sum(l2);
    s
}
        "#,
            (4 * 4 + 3 * 3 + 2 * 2 + 1) as i32,
        )
    }

    #[test]
    fn danglin_ptr() {
        test_codegen_and_run(
            r#"
enum List {
    Nil {},
    Cons {head: Int, tail: List}
}
 
fn map(l: List, f: (Int) -> Int) -> List {
    match l {
        List::Nil {} => List::Nil{},
        List::Cons{head, tail} => List::Cons{head: f(head), tail: map(tail, f)}
    }
}

fn sum(l: List) -> Int {
    match l {
        List::Nil {} => 0,
        List::Cons{head, tail} => head + sum(tail),
    }
}

fn get_list() -> _ {
    let l1 = List::Cons{head: 4, tail: List::Cons{head: 3, tail: List::Cons{head: 2, tail: List::Cons{head: 1, tail: List::Nil {}}}}};
    l1
}

fn main() -> _ {
    let l1 = get_list();
    let l2 = map(l1, (x) => x * x);
    let s = sum(l2);
    s
}
        "#,
            (4 * 4 + 3 * 3 + 2 * 2 + 1) as i32,
        )
    }
}
