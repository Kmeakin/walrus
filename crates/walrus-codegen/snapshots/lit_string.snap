---
source: crates/walrus-codegen/src/tests.rs
expression: module

---
; ModuleID = 'module'
source_filename = "module"

%String = type { i32, i8* }

@string = private unnamed_addr constant [12 x i8] c"hello world\00", align 1

define void @builtins.exit(i32 %status) {
  call void @exit(i32 %status)
  unreachable
}

declare void @exit(i32)

define {} @builtins.putchar(i32 %c) {
  %1 = call i32 @putchar(i32 %c)
  ret {} zeroinitializer
}

declare i32 @putchar(i32)

define %String @main() {
main.entry:
  ret %String { i32 11, i8* getelementptr inbounds ([12 x i8], [12 x i8]* @string, i32 0, i32 0) }
}

